### PR TITLE
Feat: add external scaler

### PR DESCRIPTION
**Pull request by chrisliu1995**
## BackGround
When stateful game servers need to be scaled horizontally, it is necessary to avoid deleting the game server that players are playing on. Therefore, the state of the GameServers need to be taken into account when scaling down. OKG determine the value of replicas according to the state of these GameServers.

## Design
OKG exposes metric gssReplicas, which indicates the number of replicas of the corresponding GameServerSet.

Only GameServers whose opsState is WaitToBeDeleted will be deleted, that is to say, the desired number of replicas is the current number of replicas minus the number of WaitToBeDeleted replicas, so that the deletion of game servers can be guaranteed to be absolutely safe.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -164,6 +169,17 @@ func main() {
 		}
 	}()
 
+	externalScaler := externalscaler.NewExternalScaler(mgr.GetClient())
+	go func() {
+		grpcServer := grpc.NewServer()
+		lis, _ := net.Listen("tcp", scaleServerAddr)
+		externalscaler.RegisterExternalScalerServer(grpcServer, externalScaler)
+		if err := grpcServer.Serve(lis); err != nil {
```
[the remark this answers is not in the document]
Comment (chrisliu1995) [reply main.go:177]: @sonatype-lift ignore

File: main.go [hunk 2]
```diff
@@ -164,6 +169,17 @@ func main() {
...
+			setupLog.Error(err, "unable to setup ExternalScalerServer")
+			os.Exit(1)
+		}
+	}()
+
 	setupLog.Info("starting kruise-game-manager")
 
 	if err := mgr.Start(signal); err != nil {
```
[the remark this answers is not in the document]
Comment (chrisliu1995) [reply main.go:185]: @sonatype-lift ignore

File: main.go [hunk 3]
```diff
@@ -164,6 +169,17 @@ func main() {
 		}
 	}()
 
+	externalScaler := externalscaler.NewExternalScaler(mgr.GetClient())
+	go func() {
+		grpcServer := grpc.NewServer()
```
[the remark this answers is not in the document]
Comment (chrisliu1995) [reply main.go:174]: @sonatype-lift ignore

File: pkg/externalscaler/externalscaler.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ExternalScaler struct {
+	client client.Client
+}
+
+func (e *ExternalScaler) mustEmbedUnimplementedExternalScalerServer() {
+	return
```
[the remark this answers is not in the document]
Comment (chrisliu1995) [reply pkg/externalscaler/externalscaler.go]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (chrisliu1995) [reply pkg/externalscaler/externalscaler.go]: @sonatype-lift ignore

**Review by ringtail — APPROVED** [review]
/LGTM

**Pull request merged**